### PR TITLE
Immediately called closures are always noescape

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes https://bugs.swift.org/browse/SR-4563, <rdar://problem/31910280> and <rdar://problem/32409133>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/attr/attr_escaping.swift [hunk 1]
```diff
@@ -160,3 +160,20 @@ class FooClass {
...
+  }()
+}
+
+class HasIVarCaptures {
+  var x: Int = 0
+
+  func method() {
+    {
+      x += 1 // no-error
+    }()
```
Comment (jrose-apple) [inline test/attr/attr_escaping.swift]: I believe the bug we were seeing showed up when the closure had an explicit argument list. Can you add a test for that?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/attr/attr_escaping.swift (hunk 1 above)
Comment (slavapestov) [inline test/attr/attr_escaping.swift]: You're right, those test cases were bogus and always passed anyway. I added more detailed tests and fixed the other bug we talked about with capture lists.

**Pull request merged**